Login: Update Networking and Yosemite layers to handle WordPress.org REST API

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #7597 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
During the login flow (unauthenticated state), we want to make requests to the WP.org REST API to handle Jetpack setup for self-hosted sites. This leads to a need for adding a new way of authentication, so this PR updates the Networking and Yosemite layers for that.

WordPress.org REST API uses [cookie authentication](https://developer.wordpress.org/rest-api/using-the-rest-api/authentication/), which has already been handled in the `WordPressKit` API. I copied the idea from [WordPressOrgRestApi](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/trunk/WordPressKit/WordPressOrgRestApi.swift) to apply the authenticator to Alamofire and adapt to the existing `Network` protocol for testability.

Other changes in the Networking layer:
- Added WordPressOrgRequest creating requests for WP.org REST API
- Added JetpackConnectionRemote and Mapper to make requests for fetching the Jetpack connection URL given a site address and authenticator.

Changes in the Yosemite layer:
- Added the new DeauthenticatedStore for handling actions in the unauthenticated state.
- Added the new JetpackConnectionAction and store for fetching jetpack connection URL
- Updated DeauthenticatedState to handle actions.

I could not add tests for JetpackConnectionStore, since the information needed to construct a network is specific to every request. Please feel free to suggest a better solution to make this testable if possible.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
The new changes haven't been integrated yet, please make sure that all unit tests pass.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add WordPressKit to Networking

itsmeichigo committed:
Add WordPressOrgAPI and JetpackConnectionRemote

itsmeichigo committed:
Add new store and action for jetpack connection

itsmeichigo committed:
Update DeauthenticatedState with JetpackConnectionStore

itsmeichigo committed:
Update podfile.lock

itsmeichigo committed:
Move WordPressOrgAPI and JetpackConnectionRemote to new group WordPressOrgRemote

itsmeichigo committed:
Remove designated initializer from Network and update other network accordingly

itsmeichigo committed:
Add WordPressOrgNetwork & Request and JetpackConnectionRemote and Mapper

itsmeichigo committed:
Make extension for Alamofire DataResponse and Result internal

itsmeichigo committed:
Update JetpackConnectionStore with the new updates of network and remote

itsmeichigo committed:
Add test json response for jetpack connection url request

itsmeichigo committed:
Add tests for JetpackConnectionURLMapper

itsmeichigo committed:
Add tests for JetpackConnectionRemote

itsmeichigo committed:
Add tests for WordPressOrgRequest

itsmeichigo committed:
Update comment for WordPressOrgNetwork

itsmeichigo commented on Networking/Networking/Network/Network.swift:
```diff
@@ -19,13 +19,6 @@ public protocol MultipartFormData {
-    /// - Parameters:
-    ///     - credentials: WordPress.com Credentials.
-    ///
-    init(credentials: Credentials)
```
This is removed to make `Network` reusable for other types. `Credentials` is now required only for `AlamofireNetwork`. In fact I think we may want to rename `AlamofireNetwork` to `JetpackTunneledNetwork`, but I'm not changing this in this PR.

itsmeichigo committed:
Keep strong reference to network and remote in JetpackConnectionStore

itsmeichigo committed:
Add validation for WordPressOrgNetwork to handle nonce retrieval

selanthiraiyan commented on Networking/Networking/Mapper/JetpackConnectionURLMapper.swift:
```diff
@@ -0,0 +1,20 @@
+
+    /// (Attempts) to convert the response into a URL.
+    ///
+    func map(response: Data) throws -> URL? {
```
What do you think about throwing an error instead of returning an optional URL? (As the method is already following that pattern) 

Also, we could follow the same pattern in other places like the completion blocks of `JetpackConnectionAction`, `JetpackConnectionRemote` and `JetpackConnectionStore`. (i.e. Send Error instead of optional URL)

itsmeichigo replied:
Good point! Updated in fc69543e61e6bf8116e3dd531f674e7960ae6f5c.

selanthiraiyan commented on Networking/NetworkingTests/Mapper/JetpackConnectionURLMapperTests.swift:
```diff
@@ -0,0 +1,26 @@
+import XCTest
+@testable import Networking
+
+/// UserMapper Unit Tests
```
Nit: `JetpackConnectionURLMapper` instead of `UserMapper`. 😄

selanthiraiyan commented on Networking/Networking/Network/WordPressOrgNetwork.swift:
```diff
@@ -0,0 +1,142 @@
+
+/// Class to handle WP.org REST API requests.
+///
+public final class WordPressOrgNetwork: Network {
```
Should we update https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/NETWORKING.md#network about this new `Network` implementation?

itsmeichigo replied:
Good thinking 💯  Updated in 14a1e2503a267fa65d4c612d5ea9863b79f1b414.

selanthiraiyan commented on WooCommerce/Classes/Yosemite/DeauthenticatedState.swift:
```diff
@@ -15,7 +26,9 @@ class DeauthenticatedState: StoresManagerState {
     func willLeave() { }
 
-    /// NO-OP: During deauth method, we're not running any actions.
+    /// During deauth method, we're not handling actions that don't require access token.
```
> we're not handling actions that don't require access token

Sorry, I don't understand this. Does this mean that we are only handling actions that require access token? 🤔

itsmeichigo replied:
This is a typo 😅  I clarified the comment in 2d6aef5aac24954d9b29a5e5b1f6b66e38931023.

selanthiraiyan commented on Yosemite/Yosemite/Base/DeauthenticatedStore.swift:
```diff
@@ -0,0 +1,47 @@
+//         Every store is subscribed to the global action dispatcher (although it can be initialized with a custom dispatcher), and should
+//         respond to relevant Actions by implementing onAction(_:), and change its internal state according to those actions.
+//
+open class DeauthenticatedStore: ActionsProcessor {
```
❓  Could this be `public` instead of `open`?

selanthiraiyan commented on Networking/Networking/Requests/WordPressOrgRequest.swift:
```diff
@@ -0,0 +1,54 @@
+
+/// Represents a WordPress.org REST API Endpoint
+///
+struct WordPressOrgRequest: URLRequestConvertible {
```
I think we need to update https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/NETWORKING.md#urlrequestconvertible about the new `URLRequestConvertible` implementation. 😄

itsmeichigo replied:
Also updated in 2d6aef5aac24954d9b29a5e5b1f6b66e38931023.

selanthiraiyan commented on Networking/Networking/Requests/WordPressOrgRequest.swift:
```diff
@@ -0,0 +1,54 @@
+
+    /// Parameters
+    ///
+    let parameters: [String: Any]
```
```suggestion
    var parameters: [String: Any] = [:]
```

Can we do this to avoid having a custom `init` method? 🤔

itsmeichigo replied:
Makes sense! Updated in 788471b65b71cc00598046029bea2b523340fa44.

selanthiraiyan commented on Networking/Networking/Requests/WordPressOrgRequest.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+
+
+    /// Returns a URLRequest instance reprensenting the current Jetpack Request.
```
Nit: The comment needs to be updated.

selanthiraiyan commented on Networking/Networking/Network/AlamofireNetwork.swift:
```diff
@@ -109,7 +109,7 @@ public class AlamofireNetwork: Network {
 
 /// MARK: - Alamofire.DataResponse: Private Methods
```
Nit: We should also update this comment.

selanthiraiyan commented on Yosemite/Yosemite/Stores/JetpackConnectionStore.swift:
```diff
@@ -0,0 +1,48 @@
+public final class JetpackConnectionStore: DeauthenticatedStore {
+
+    // Keep a strong reference to network to keep requests alive
+    private var network: WordPressOrgNetwork?
```
❓  I hope that `Remote` will [hold a reference](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Networking/Networking/Remote/Remote.swift#L11) to `Network`. And we are holding a strong reference to `JetpackConnectionRemote` down below. 
Do we still have to hold a strong reference to `WordPressOrgNetwork`?

itsmeichigo replied:
You are right - there's no need to keep the reference to `network` here. Some of our other stores seem to keep strong references to both `network` and `remote` as `network` is a requirement for `Store`. Do you think this can cause retained cycles?

Anyway - I updated `DeauthenticatedStore` following this suggestion in 0a61d25a33c157d2870ca0eb6d9dddeab16575b5.

selanthiraiyan commented on Networking/Networking/Network/WordPressOrgNetwork.swift:
```diff
@@ -0,0 +1,142 @@
+    ///     - request: Request that should be performed.
+    ///     - completion: Closure to be executed upon completion.
+    ///
+    /// - Note:
+    ///     - The response body will always be returned (when possible), even when there's a networking error.
+    ///       This differs slightly from the standard Alamofire `.validate()` behavior, and it's required so that
+    ///       the upper layers can properly detect "Jetpack Tunnel" Errors.
+    ///     - Yes. We do the above because the Jetpack Tunnel endpoint doesn't properly relay the correct statusCode.
+    ///
```
> Yes. We do the above because the Jetpack Tunnel endpoint doesn't properly relay the correct statusCode.

❓ Is this comment valid here?

itsmeichigo replied:
Updated comments for `WordPressOrgNetwork` in 0c5b2374329363b44cb0b476b43e9d76f1652759.

selanthiraiyan reviewed (approved):
Great work, Huong. 🎉 

I left a few non-blocking nits and questions. 

LGTM :shipit:

itsmeichigo committed:
Throw error instead of return nil when mapping connection url

itsmeichigo committed:
Update comment for JetpackConnectionURLMapperTests

itsmeichigo committed:
Update Networking.md for new changes

itsmeichigo committed:
Fix typo in DeauthenticatedState comment

itsmeichigo committed:
Make DeauthenticatedStore public instead of open

itsmeichigo committed:
Make parameters optional in WordPressOrgRequest

itsmeichigo committed:
Fix typo in comment for WordPressOrgRequest

itsmeichigo committed:
Update comment for Alamofire extension

itsmeichigo committed:
Update comments on WordPressOrgNetwork

itsmeichigo committed:
Only keep strong reference of remote in JetpackConnectionStore

itsmeichigo committed:
Merge branch 'trunk' into issue/7597-network-yosemite-wporg-api

itsmeichigo committed:
Update podfile.lock

itsmeichigo committed:
Silence hound

itsmeichigo commented:
Thank you @selanthiraiyan for the review 🙇 I've addressed the issues you found, please feel free to double-check and let me know if anything else needs to be updated.

itsmeichigo committed:
Add new models for Jetpack connection data

itsmeichigo committed:
Add mapper for JetpackConnectedUser

itsmeichigo committed:
Add test response for JetpackConnectedUser

itsmeichigo committed:
Add tests for JetpackConnectedUserMapper

itsmeichigo committed:
Rename JetpackConnectedUser to JetpackUser

itsmeichigo committed:
Add new endpoint for fetching Jetpack user

itsmeichigo committed:
Update tests for JetpackConnectionRemote

itsmeichigo committed:
Add new action and store for fetching Jetpack user

itsmeichigo committed:
Update deauthenticated store and state to make deauthenticated store testable

itsmeichigo committed:
Add tests for JetpackConnectionStore

itsmeichigo committed:
Update typo in comment

itsmeichigo committed:
Generate copiable for JetpackUser and WordPressComUser

itsmeichigo committed:
Rename WordPressComUser to DotComUser

itsmeichigo committed:
Use default decoding logic

itsmeichigo committed:
Remove unused properties for DotComUser

itsmeichigo committed:
Rename jetpackConnectionUser endpoint and remove unused error

itsmeichigo committed:
Rename test to match the tested subject

itsmeichigo committed:
Update comment explaining the purpose of updateDeauthenticatedState

itsmeichigo committed:
Update comments on JetpackConnectionAction

itsmeichigo committed:
Update JetpackUserMapperTests

itsmeichigo committed:
Rename DotComUser to DotcomUser

itsmeichigo committed:
Restore custom decoding initializer for JetpackUser to fix test failure

itsmeichigo committed:
Remove updateDeauthenticatedState

itsmeichigo committed:
Remove updateStores from DeauthenticatedState

itsmeichigo committed:
Add action to update remote for JetpackConnectionAction

itsmeichigo committed:
Update unit tests for JetpackConnectionStore

itsmeichigo committed:
Remove white line

itsmeichigo committed:
Rename test in JetpackConnectionStoreTests

itsmeichigo committed:
Add typealias for DotComUser and JetpackUser

itsmeichigo committed:
Remove redundant imports

itsmeichigo committed:
Merge pull request #7663 from woocommerce/issue/7597-network-yosemite-jetpack-data

Login: Update Networking and Yosemite layers to handle Jetpack connection data API

itsmeichigo committed:
Merge branch 'trunk' into issue/7597-network-yosemite-wporg-api

itsmeichigo committed:
Update podfile.lock
